OrbitGl/App.cpp

```diff
@@ -1348,3 +1348,55 @@ void OrbitApp::DeselectTracepoint(const TracepointInfo& tracepoint) {
 [[nodiscard]] bool OrbitApp::IsTracepointSelected(const TracepointInfo& info) const {
   return data_manager_->IsTracepointSelected(info);
 }
+
+void OrbitApp::AddFrameTrack(const orbit_client_protos::FunctionInfo& function) {
+  const CaptureData& capture_data = GetCaptureData();
+  auto function_address = capture_data.GetAbsoluteAddress(function);
+  std::vector<std::shared_ptr<TimerChain>> chains =
+      GCurrentTimeGraph->GetAllThreadTrackTimerChains();
+
+  std::vector<uint64_t> all_start_times;
+
+  for (auto& chain : chains) {
+    if (!chain) continue;
+    for (TimerChainIterator it = chain->begin(); it != chain->end(); ++it) {
+      TimerBlock& block = *it;
+      for (uint64_t i = 0; i < block.size(); i++) {
+        TextBox& box = block[i];
+        if (box.GetTimerInfo().function_address() == function_address) {
+          all_start_times.push_back(box.GetTimerInfo().start());
+        }
+      }
+    }
+  }
+  std::sort(all_start_times.begin(), all_start_times.end());
+
+  for (size_t k = 0; k < all_start_times.size() - 1; ++k) {
+    TimerInfo frame_timer;
+
+    // TID is meaningless for this timer (start and end can be on two different threads).
+    constexpr const int32_t kUnusedThreadId = -1;
+    frame_timer.set_thread_id(kUnusedThreadId);
+    frame_timer.set_start(all_start_times[k]);
+    frame_timer.set_end(all_start_times[k + 1]);
+    // We use user_data_key to keep track of the frame number.
+    frame_timer.set_user_data_key(k);
+    frame_timer.set_type(TimerInfo::kFrame);
+
+    GCurrentTimeGraph->ProcessTimer(frame_timer, &function);
+  }
+  frame_track_functions_.insert(std::make_pair(function_address, function));
```

Comment:
I'm wondering if we should early out in the case where frame_track_functions_ already has an entry for function_address, does it make sense to call AddFrameTrack twice with the same function?

Reply:
The logic that calls this function already checks if a frame track has been added for this function, but it might be a good idea to check that here as well and early out. If this is run twice, it just adds a duplicate timer to the existing frame track for this function. I don't think we dedup, so this would increase runtime for UpdatePrimitives.

I wasn't sure whether to silently return, add an error or even a CHECK. I've now added an ERROR log message and DCHECK based on the following reasoning:

- If in a shipped version this function is called twice due to a bug, it's not really a big deal. Crashing and making this feature unusable would be worse. Hence I don't want to have a CHECK here.
- It's still a logic mistake that we should fix and to draw attention to it, the combination of ERROR and DCHECK seems to be reasonable. 

I added a similar check to RemoveFrameTrack, which should only be called on functions that have a frametrack.

Does this sound reasonable?

Reply:
Sounds reasonable to me!

Reply:
Ok, thanks!

Reply:
We discussed a little in the daily about this decision. I'm not sure yet which is the best, but sounds reasonable!